src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs

```diff
@@ -336,6 +372,176 @@ protected virtual unsafe IntPtr AppWndProc(IntPtr hWnd, uint msg, IntPtr wParam,
+                    {
+                        //notifies about proximity of pointer device to the digitizer.
+                        //contains pointer id and proximity.
+                        //https://docs.microsoft.com/en-us/previous-versions/windows/desktop/inputmsg/wm-pointerdeviceinrange
+                        break;
+                    }
+                case WindowsMessage.WM_NCPOINTERUPDATE:
+                    {
+                        //NC stands for non-client area - window header and window border
+                        //As I found above in an old message handling - we dont need to handle NC pointer move/updates.
+                        //All we need is pointer down and up. So this is skipped for now.
+                        break;
+                    }
+                case WindowsMessage.WM_NCPOINTERDOWN:
+                case WindowsMessage.WM_NCPOINTERUP:
+                case WindowsMessage.WM_POINTERDOWN:
+                case WindowsMessage.WM_POINTERUP:
+                case WindowsMessage.WM_POINTERUPDATE:
+                    {
+                        GetDevicePointerInfo(wParam, out var device, out var info, ref timestamp);
+
+                        if (info.historyCount > 1)
+                        {
+                            if (info.pointerType == PointerInputType.PT_TOUCH)
+                            {
+                                if (ShouldIgnoreTouchEmulatedMessage())
+                                {
+                                    break;
+                                }
+
+                                var pointerId = (uint)(ToInt32(wParam) & 0xFFFF);
+                                var historyCount = (int)info.historyCount;
+                                var historyTouchInfos = new POINTER_TOUCH_INFO[historyCount];
+                                if (GetPointerTouchInfoHistory(pointerId, ref historyCount, historyTouchInfos))
+                                {
+                                    //last info is the same as the current so skip it
+                                    for (int i = 0; i < historyCount - 1; i++)
+                                    {
+                                        var historyTouchInfo = historyTouchInfos[i];
+                                        var historyInfo = historyTouchInfo.pointerInfo;
+                                        var historyEventType = GetEventType(message, historyInfo);
+                                        var historyPoint = PointToClient(new PixelPoint(
+                                            historyInfo.ptPixelLocationX, historyInfo.ptPixelLocationY));
+                                        var historyModifiers = GetInputModifiers(historyInfo.dwKeyStates);
+                                        var historyTimestamp = historyInfo.dwTime == 0 ? timestamp : historyInfo.dwTime;
+                                        Input?.Invoke(new RawTouchEventArgs(_touchDevice, historyTimestamp, _owner,
```

Comment:
Instead of invoking each event separately, you should use intermediate points array.
See "RawPointerEventArgs.IntermediatePoints". Although its type should be probably changed from "Point[]" to something else that will also contain properties.

Reply:
You've added this yesterday... ok.

Reply:
IntermediatePoints - is not a best solution because when using touch or pen - every intermediate point can have a custom pressure/tilt/twist. So I think we can keep it like this now and can refactor later.

Reply:
>every intermediate point can have a custom pressure/tilt/twist

That's exactly the point: instead of spamming the app with tons of events we group everything into one.